File: csharp/Vec3.cs
using System;
using System.Diagnostics;

namespace ToyRayTrace
{
    public readonly struct Vec3
    {
        static readonly Vec3 k_Zero = new Vec3(0, 0, 0);
        static readonly Vec3 k_One = new Vec3(1, 1, 1);
        public static ref readonly Vec3 Zero => ref k_Zero;
        public static ref readonly Vec3 One => ref k_One;

        public readonly float x, y, z;

        public Vec3(float v1, float v2, float v3)
        {
            x = v1;
            y = v2;
            z = v3;
        }

        public Vec3(in Vec3 v)
        {
            x = v.x;
            y = v.y;
            z = v.z;
        }

        public float Length => MathF.Sqrt(x * x + y * y + z * z);
        public float SquaredLength => x * x + y * y + z * z;
        public bool IsNormalized => Math.Abs(SquaredLength - 1.0f) < 0.01f;

        public static Vec3 operator +(in Vec3 v) => new Vec3(v);
        public static Vec3 operator -(in Vec3 v) => new Vec3(-v.x, -v.y, -v.z);
        public static Vec3 operator +(in Vec3 a, in Vec3 b) => new Vec3(a.x + b.x, a.y + b.y, a.z + b.z);
        public static Vec3 operator -(in Vec3 a, in Vec3 b) => new Vec3(a.x - b.x, a.y - b.y, a.z - b.z);
        public static Vec3 operator *(in Vec3 a, in Vec3 b) => new Vec3(a.x * b.x, a.y * b.y, a.z * b.z);
        public static Vec3 operator /(in Vec3 a, in Vec3 b) => new Vec3(a.x / b.x, a.y / b.y, a.z / b.z);
        public static Vec3 operator *(in Vec3 v, float f) => new Vec3(v.x * f, v.y * f, v.z * f);
        public static Vec3 operator /(in Vec3 v, float f) => new Vec3(v.x / f, v.y / f, v.z / f);

        public static Vec3 Normalize(in Vec3 v)
        {
            var k = 1 / v.Length;
            return new Vec3(v.x*k, v.y*k, v.z*k);
        }

        public static float Dot(in Vec3 a, in Vec3 b) => a.x * b.x + a.y * b.y + a.z * b.z;

        public static Vec3 Cross(in Vec3 a, in Vec3 b)
        {
            return new Vec3(
                a.y*b.z - a.z*b.y,
                -(a.x*b.z - a.z*b.x),
                a.x*b.y - a.y*b.x);
        }

        public static Vec3 Reflect(in Vec3 v, in Vec3 normal)
        {
            Debug.Assert(v.IsNormalized);
            return v - normal * 2 * Dot(v, normal);
        }

        public static bool Refract(in Vec3 v, in Vec3 normal, float niOverNt, out Vec3 refracted)
        {
            Debug.Assert(v.IsNormalized);
            var dt = Dot(v, normal);
            var discriminant = 1.0f - niOverNt * niOverNt * (1 - dt * dt);
            if (discriminant > 0)
            {
                refracted = (v - normal* dt) * niOverNt - normal * MathF.Sqrt(discriminant);
                Debug.Assert(refracted.IsNormalized);
                return true;
            }
            refracted = Zero;
            return false;
        }
    }
}

File: rust/src/raytrace/vec3.rs
use rand::Rng;
use std::ops;

#[derive(Copy, Clone, Default)]
pub struct Vec3 {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}

impl Vec3 {
    pub const ZERO: Vec3 = Vec3 { x: 0.0, y: 0.0, z: 0.0 };
    pub const ONE: Vec3 = Vec3 { x: 1.0, y: 1.0, z: 1.0 };
    pub const Y: Vec3 = Vec3 { x: 0.0, y: 1.0, z: 0.0 };

    pub fn new(x: f32, y: f32, z: f32) -> Self {
        Vec3 { x: x, y: y, z: z }
    }

    pub fn length(&self) -> f32 {
        (self.x * self.x + self.y * self.y + self.z * self.z).sqrt()
    }

    pub fn squared_length(&self) -> f32 {
        self.x * self.x + self.y * self.y + self.z + self.z
    }

    pub fn is_normalized(&self) -> bool {
        (self.squared_length() - 1.0).abs() < 0.001
    }
}

impl ops::Neg for Vec3 {
    type Output = Vec3;
    fn neg(self) -> Vec3 {
        Vec3 {
            x: -self.x,
            y: -self.y,
            z: -self.z,
        }
    }
}

impl ops::Add<Vec3> for Vec3 {
    type Output = Vec3;
    fn add(self, rhs: Vec3) -> Vec3 {
        Vec3 {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
            z: self.z + rhs.z,
        }
    }
}

impl ops::Sub<Vec3> for Vec3 {
    type Output = Vec3;
    fn sub(self, rhs: Vec3) -> Vec3 {
        Vec3 {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
            z: self.z - rhs.z,
        }
    }
}

impl ops::Mul<Vec3> for Vec3 {
    type Output = Vec3;
    fn mul(self, rhs: Vec3) -> Vec3 {
        Vec3 {
            x: self.x * rhs.x,
            y: self.y * rhs.y,
            z: self.z * rhs.z,
        }
    }
}

impl ops::Mul<f32> for Vec3 {
    type Output = Vec3;
    fn mul(self, rhs: f32) -> Vec3 {
        Vec3 {
            x: self.x * rhs,
            y: self.y * rhs,
            z: self.z * rhs,
        }
    }
}

impl ops::Div<Vec3> for Vec3 {
    type Output = Vec3;
    fn div(self, rhs: Vec3) -> Vec3 {
        Vec3 {
            x: self.x / rhs.x,
            y: self.y / rhs.y,
            z: self.z / rhs.z,
        }
    }
}

impl ops::Div<f32> for Vec3 {
    type Output = Vec3;
    fn div(self, rhs: f32) -> Vec3 {
        Vec3 {
            x: self.x / rhs,
            y: self.y / rhs,
            z: self.z / rhs,
        }
    }
}

pub fn dot(a: Vec3, b: Vec3) -> f32 {
    a.x * b.x + a.y * b.y + a.z * b.z
}

pub fn cross(a: Vec3, b: Vec3) -> Vec3 {
    Vec3 {
        x: a.y * b.z - a.z * b.y,
        y: -(a.x * b.z - a.z * b.x),
        z: a.x * b.y - a.y * b.x,
    }
}

pub fn normalize(v: Vec3) -> Vec3 {
    let inv_len = 1.0 / v.length();
    Vec3 {
        x: v.x * inv_len,
        y: v.y * inv_len,
        z: v.z * inv_len,
    }
}

pub fn reflect(v: Vec3, n: Vec3) -> Vec3 {
    v - n * (2.0 * dot(v, n))
}

pub fn refract(v: Vec3, n: Vec3, ni_over_nt: f32) -> Option<Vec3> {
    let dt = dot(v, n);
    let discriminant = 1.0 - ni_over_nt * ni_over_nt * (1.0 - dt * dt);
    if discriminant > 0.0 {
        Some((v - n * dt) * ni_over_nt - n * discriminant.sqrt())
    } else {
        None
    }
}

pub fn random_in_unit_sphere() -> Vec3 {
    let mut p = Vec3::ONE;
    let mut rng = rand::thread_rng();
    while p.squared_length() > 1.0 {
        p = Vec3::new(rng.gen(), rng.gen(), rng.gen()) * 2.0 - Vec3::ONE;
    }
    p
}

pub fn random_in_unit_disc() -> Vec3 {
    let mut p = Vec3::ONE;
    let mut rng = rand::thread_rng();
    while p.squared_length() > 1.0 {
        p = Vec3::new(rng.gen(), rng.gen(), 0.0) * 2.0 - Vec3::ONE;
    }
    p
}
